[InstCombine] SimplifyDemandedBits: add basic bitcast handling, Fixes #173780

ayushpareek2003 opened the pull request:
this patch teaches InstCombine's SimplifyDemandedBits to peek through lossless bitcasts with identical scalar bit widths. This avoids the need for target-specific handling of bitcasted masks (e.g. X86 blend intrinsics).
Fixes #173780

ayushpareek2003 committed:
Enhance SimplifyDemandedBits with bitcast handling

Add support for simplifying demanded bits through lossless bitcasts.

ayushpareek2003 committed:
refactored simplifydemandedBits logic for Mask handling

dtcxzyw reviewed (commented):
Missing tests

ayushpareek2003 committed:
Merge branch 'main' into main

ayushpareek2003 committed:
test for SimplifyDemandedBits bitcast handling

ayushpareek2003 commented:
> Missing tests
The added test fails without this change because SimplifyDemandedBits stops at bitcast instructions and cannot reason about the sign bit. With the patch applied, demanded-bit analysis correctly peeks through same-width bitcasts, enabling full constant folding

nikic commented:
@ayushpareek2003 The test has the same behavior without your change: https://llvm.godbolt.org/z/hfoqf7cv3

RKSimon commented:
Try tests with a single bitcast from v1i32 to i32 instead, double bitcasts will fold away

ayushpareek2003 committed:
Update test for SimplifyDemandedBits with bitcasts

Updated the test to use a single v1i32 → i32 bitcast as suggested
This avoids double-bitcast canonicalization and ensures the demanded-bits analysis exercises the bitcast-handling logic added in this patch

ayushh-1024 committed:
done

ayushpareek2003 committed:
Merge branch 'llvm:main' into main

ayushh-1024 committed:
reverted unwanted changes

ak1932 commented on llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:
```diff
@@ -175,6 +175,22 @@ bool InstCombinerImpl::SimplifyDemandedBits(Instruction *I, unsigned OpNo,
+                                  Depth + 1);
+    }
+  }
+
```
I think this should be after max recursion depth check

ayushpareek2003 replied:
yes, fixed this in my latest commit

ak1932 commented on utils/bazel/.bazelignore:
```diff
@@ -1,2 +1,2 @@
 llvm-project-overlay
-examples
+examples
```
This change can be discarded

RKSimon commented on llvm/test/Transforms/InstCombine/simplify-demanded-bits-bitcast.ll:
```diff
@@ -0,0 +1,15 @@
+  ; Only the sign bit is demanded here
+  %cmp = icmp slt i32 %bc, 0
+  ret i1 %cmp
+}
```
This test still doesn't need the SimplifyDemandedBits code afaict

ayushh-1024 committed:
fixed formating and small fix in logic

Signed-off-by: ayush-1024 <ayush@kavida.ai>

RKSimon committed:
Merge branch 'main' into main
